### DISCUSSION TITLE
Need a help with TACSBuilder in MPhys

### DISCUSSION THREAD

**Discussion by MohdAhmet** [OpenMDAO/mphys#185] [Q&A] [answered]
I am working on a structural optimization project for a grillage model (bulk carrier hatch cover) using TACS with OpenMDAO. The objective is to minimize mass while satisfying yield stress and deflection constraints. I closely followed the example scripts provided in the TACS documentation as follows:
Initialize TACS builder

class PlateModel(Multipoint):
def setup(self):
struct_builder = TacsBuilder(
mesh_file=bdf_file,
assembler_setup=None,
element_callback=elem_callback,
problem_setup=problem_setup
)
struct_builder.initialize(self.comm)

but encountered the following error during execution:

Opt.py", line 62, in
struct_builder = TacsBuilder(
TypeError: init() got an unexpected keyword argument 'mesh_file'
mdolabuser@d007da2d526e:~/repos/tacs/examples/SCB/GrillageOpt$

My model is fully defined in the .dat file and verified independantly as well before trying to run the optimization.
The assembler is initilaized without arguments for this reason.

could you please suggest what is the best course of action? I have been reseaching for one week in the documentation to solve this error but I couldn't

**Comment by timryanb** [comment, marked as answer]
Hi @MohdAhmet it sounds like you're using an out of date TACS version, please update to version 3.5.0 or higher.

**Reply by MohdAhmet** [reply]
Thanks for your answer, I have TACS installed in Docker linux, any clue how to update it?

Thanks in advance

**Reply by timryanb** [reply]
Is this a docker from UMich/MDOLab?

**Reply by MohdAhmet** [reply]
mdolab yes

**Reply by timryanb** [reply]
You'll probably need to pull a newer docker build from them. Tagging some UMich folks for follow up @lamkina @anilyil

**Reply by timryanb** [reply]
It actually sounds like the MDOLab Docker's TACS build is very out of date. If you only require TACS for this optimization, I would recommend installing a newer version through miniconda. There's instructions for how to do this on the README of the TACS [repo](https://github.com/smdogroup/tacs). You'll need to manually pip install mphys in this case.

**Reply by MohdAhmet** [reply]
I actually installed TACS using MDOlab docker because I faced many problems using the miniconda instructions. 
I will try again installing it, but do you advice cleaning files or libraries before doing so?

**Reply by timryanb** [reply]
No, it shouldn't require any library cleaning

**Reply by MohdAhmet** [reply]
Ok, thanks a lot, I will try to install it again, would you mind to continue this discussion in case I faced any issues?

**Reply by timryanb** [reply]
Yes, if you have a specific issue with the anaconda install I'd like to hear it. It'd be great if you could post it to the issues page on the TACS repo for visibility

**Reply by timryanb** [reply]
Worst-case scenario you may be able to use the older version of TACS packaged in the Docker for a little while. If you want an example of how to use MPhys with the version of TACS packaged in the docker you can look at this [example](https://github.com/OpenMDAO/mphys/blob/b3b4f1f5eb7f3ba0069b1742eaf851a39a525534/examples/struct_only/crm9/run_tacs_mphys.py).

Though long-term I would still recommend finding a way to get an updated tacs as we no longer support versions <3.5.0

**Reply by A-CGray** [reply]
Hi @MohdAhmet , like Tim says, the TACS version in our docker images is pretty out of date at the moment. I just opened a PR in our internal repo to update the TACS version the images use, if all goes well we could have new images out in a few days, but I can't guarantee that.

In the meantime, you could try checking out the latest version of TACS and rebuilding it, to do this you can go into `~/repos/tacs` and run:

```bash
git checkout v3.7.2
make clean
make
make interface
```

You might have to setup your git credentials in the docker image to be able to do the checkout.

**Reply by MohdAhmet** [reply]
Ok, this is nice, I will try my best installing the newest version and let you know in case of any complications in the issues section.
Thanks a lot for your help

**Reply by A-CGray** [reply]
@MohdAhmet , we updated all our docker images to run TACS v3.7.2, the new images get built and pushed to dockerhub overnight so they should be available tomorrow.

**Reply by MohdAhmet** [reply]
That's really nice, thank you so much, I'll try to pull the new image tommorow

**Reply by MohdAhmet** [reply]
I have succesfully pulled the new image, thank you very much for your cooperation, Truely appreciated
